ports/nrf/modules/machine/adc.c

```diff
@@ -30,80 +30,120 @@
-    {{&machine_adc_type}, &ADCHandle1},
-    {{&machine_adc_type}, &ADCHandle2},
-    {{&machine_adc_type}, &ADCHandle3},
-    {{&machine_adc_type}, &ADCHandle4},
-    {{&machine_adc_type}, &ADCHandle5},
-    {{&machine_adc_type}, &ADCHandle6},
-    {{&machine_adc_type}, &ADCHandle7},
+STATIC machine_adc_obj_t machine_adc_obj[] = {
+#if NRF51
+    {{&machine_adc_type}, .id = 0, .ain = NRF_ADC_CONFIG_INPUT_0},
+    {{&machine_adc_type}, .id = 1, .ain = NRF_ADC_CONFIG_INPUT_1},
+    {{&machine_adc_type}, .id = 2, .ain = NRF_ADC_CONFIG_INPUT_2},
+    {{&machine_adc_type}, .id = 3, .ain = NRF_ADC_CONFIG_INPUT_3},
+    {{&machine_adc_type}, .id = 4, .ain = NRF_ADC_CONFIG_INPUT_4},
+    {{&machine_adc_type}, .id = 5, .ain = NRF_ADC_CONFIG_INPUT_5},
+    {{&machine_adc_type}, .id = 6, .ain = NRF_ADC_CONFIG_INPUT_6},
+    {{&machine_adc_type}, .id = 7, .ain = NRF_ADC_CONFIG_INPUT_7},
+#else
+    {{&machine_adc_type}, .id = 0},
+    {{&machine_adc_type}, .id = 1},
+    {{&machine_adc_type}, .id = 2},
+    {{&machine_adc_type}, .id = 3},
+    {{&machine_adc_type}, .id = 4},
+    {{&machine_adc_type}, .id = 5},
+    {{&machine_adc_type}, .id = 6},
+    {{&machine_adc_type}, .id = 7},
+#endif
 };
 
+#if defined(NRF52_SERIES)
+STATIC void saadc_event_handler(nrfx_saadc_evt_t const * p_event) {
+    (void)p_event;
+}
+#endif
+
+void adc_init0(void) {
+#if defined(NRF52_SERIES)
+    const nrfx_saadc_config_t config = {
+        .resolution         = NRF_SAADC_RESOLUTION_8BIT,
+        .oversample         = NRF_SAADC_OVERSAMPLE_DISABLED,
+        .interrupt_priority = 6,
+        .low_power_mode     = false
+    };
+
+    nrfx_saadc_init(&config, saadc_event_handler);
+#endif
+}
+
 STATIC int adc_find(mp_obj_t id) {
     // given an integer id
     int adc_id = mp_obj_get_int(id);
 
     int adc_idx = adc_id;
 
     if (adc_idx >= 0 && adc_idx <= MP_ARRAY_SIZE(machine_adc_obj)
-        && machine_adc_obj[adc_idx].adc != NULL) {
+        && machine_adc_obj[adc_idx].id != -1) {
         return adc_idx;
     }
     nlr_raise(mp_obj_new_exception_msg_varg(&mp_type_ValueError,
         "ADC(%d) does not exist", adc_id));
 }
 
-
 /// \method __str__()
 /// Return a string describing the ADC object.
 STATIC void machine_adc_print(const mp_print_t *print, mp_obj_t o, mp_print_kind_t kind) {
     machine_adc_obj_t *self = o;
 
-    (void)self;
-
-    mp_printf(print, "ADC()");
+    mp_printf(print, "ADC(%u)", self->id);
 }
 
 /******************************************************************************/
 /* MicroPython bindings for machine API                                       */
 
 // for make_new
-enum {
-    ARG_NEW_PIN,
-};
-
 STATIC mp_obj_t machine_adc_make_new(const mp_obj_type_t *type, size_t n_args, size_t n_kw, const mp_obj_t *all_args) {
+    enum { ARG_pin };
     static const mp_arg_t allowed_args[] = {
-        { ARG_NEW_PIN,  MP_ARG_OBJ, {.u_obj = MP_OBJ_NEW_SMALL_INT(-1) } },
+        { ARG_pin,  MP_ARG_OBJ, {.u_obj = MP_OBJ_NEW_SMALL_INT(-1) } },
```

Comment:
I think this should be `MP_QSTR_pin`.

Reply:
This is even wrong as i can see it. For now, this is really the instance of the AIN. Selecting AIN (ADC instance id) is not yet implemented. So, i suggest to rename this to "id" instead.